tools/mac/bundle_mac_dylibs.sh

```diff
@@ -0,0 +1,119 @@
+IGNORE_LIB_PREFIX=("/usr/lib/" "/System/" "@executable_path*")
+
+# State
+
+LIBRARY_PATHS=()
+COPY_ONLY=0
+LINKS_ONLY=0
+
+# Business logic
+
+print_help() {
+    echo "Usage: $0 [--copy-only] [--links-only] [--help]"
+    echo "Copies shared libraries into the macOS app bundle for $APP_NAME."
+    echo "  --copy-only     Only copy libraries, do not update links."
+    echo "  --links-only    Only update links, do not copy libraries."
+    echo "  --help          Display this help and exit."
+}
+
+for arg in "$@"; do
+    case $arg in
+        --copy-only)
+            COPY_ONLY=1
+            ;;
+        --links-only)
+            LINKS_ONLY=1
+            ;;
+        --help)
+            print_help
+            exit 0
+            ;;
+    esac
+done
+
+echo "Gathering libraries for $APP_NAME ..."
+
+should_ignore_lib() {
+    local lib_path=$1
+    for lib_prefix in "${IGNORE_LIB_PREFIX[@]}"; do
+        [[ $lib_path == $lib_prefix* ]] && return 0
+    done
+    return 1
+}
+
+gather_libs() {
+    local binary_path=$1
+    local parent_path=$(dirname "$binary_path")
+    local libs=$(otool -L "$binary_path" | awk 'NR>1 {print $1}')
+
+    for lib in $libs; do
+        if [[ $lib == @rpath/* ]]; then
+            lib="${parent_path}/${lib#@rpath/}"
+        elif [[ $lib != /* ]]; then
+            lib="$parent_path/$lib"
+        fi
+
+        if should_ignore_lib "$lib" \
+            || [[ $lib == /tmp* ]] \
+            || [[ " ${LIBRARY_PATHS[*]} " == *" $lib "* ]] \
+            || [[ $lib == $binary_path ]]; then
+            continue
+        fi
+
+        LIBRARY_PATHS+=("$lib")
+        gather_libs "$lib"
+    done
+}
+
+copy_libs() {
+    for lib in "${LIBRARY_PATHS[@]}"; do
+        local lib_name=$(basename "$lib")
+        echo "Copying $lib to $FRAMEWORKS_PATH"
+        cp -n "$lib" "$FRAMEWORKS_PATH/$lib_name"
+    done
+}
+
+update_links() {
+    local binary_path=$1
+    local libs=$(otool -L "$binary_path" | awk 'NR>1 {print $1}')
+
+    for lib in $libs; do
+        if should_ignore_lib "$lib"; then
+            continue
+        fi
+
+        local lib_name=$(basename "$lib")
+        echo "Updating link for $lib_name in $binary_path"
+        install_name_tool -change "$lib" "@executable_path/../Frameworks/$lib_name" "$binary_path"
+        
+        if [[ $lib_name == $(basename "$binary_path") ]]; then
+            echo "Updating id for $lib_name in $binary_path"
+            install_name_tool -id "@executable_path/../Frameworks/$lib_name" "$binary_path"
+        fi
+    done
+}
+
+if [[ $COPY_ONLY -eq 1 ]] || [[ $COPY_ONLY -eq 0 && $LINKS_ONLY -eq 0 ]]; then
+    gather_libs "$APP_BINARY_PATH"
+    mkdir -p "$FRAMEWORKS_PATH"
+
+    copy_libs
+fi
+
+if [[ $LINKS_ONLY -eq 1 ]] || [[ $COPY_ONLY -eq 0 && $LINKS_ONLY -eq 0 ]]; then
+    for lib_path in "$FRAMEWORKS_PATH"/*; do
+        update_links "$lib_path"
+    done
+
+    update_links "$APP_BINARY_PATH"
+fi
+
+echo "Libraries for $APP_NAME copied and updated."
```

Comment:
For more complex scripts, we try to stay away from Bash and Shell – can we convert this into a Python script using `argparse`?